docs/2019-04-23-Applicative-Kore.md

```diff
@@ -1,222 +1,286 @@
+Another noticeable fact about `\exists` is its _binding behavior_; that is, it binds all free occurrences of the variable `X` in `P`. Binding behavior is important when using _substitution_, which is also defined as a notation and we will discuss about it later. In short, substitution is the so-called _capture-avoiding-substitution_, meaning that bound variables are renamed automatically to prevent unintended variable capture. The binding behavior of `\exists X P` means that its first argument `X` may be renamed in substitution. Finally, note that we put the `binder` attribute along with the declaration, so that the backends do not need to reverse engineer the information from the declaration. 
 
+#### Notations with `assoc` attribute
+
+Some _binary notations_ are _associative_. One typical example is `\or`. It is inconvenient to write a disjunction of multiple patterns using the binary notation `\or`. Instead, we want to abbreviate `\or (\or (\or a b) c) d` as `\or a b c d`. This is achieved by using the `assoc` attribute in notation declarations. 
+
+Specifically speaking, if `foo` is a binary notation, then we can use the `left-assoc` (or `right-assoc`) to specify that `foo` is left (or right) associative. Therefore,
+
+|                   | [left-assoc]                  | [right-assoc]                 |
+| ----------------- | ----------------------------- | ----------------------------- |
+| `foo a1 a2`       | `foo a1 a2`                   | `foo a1 a2`                   |
+| `foo a1 a2 a3`    | `foo (foo a1 a2) a3`          | `foo a1 (foo a2 a3)`          |
+| `foo a1 a2 a3 a4` | `foo (foo (foo a1 a2) a3) a4` | `foo a1 (foo a2 (foo a3 a4))` |
+| etc.              | etc.                          | etc.                          |
+
+_Remark._ The attributes `left-assoc` and `right-assoc` do not carry semantic meaning. They carry only syntactic meaning. Specifically speaking, they tell the _verifier_ to accept patterns of the form `foo a1 a2 a3`, even if `foo` has been declared to have arity 2. Without these attributes, the verifier will reject the pattern `foo a1 a2 a3` because it is not wellformed.
+
+As an example, here is the declaration of `\or` using `left-assoc` attribute:
+
+```
+notation \or P Q = \implies (\neg P) Q [left-assoc]
 ```
-notation \and P Q = \neg (\or (\neg P) (\neg Q))
-left-assoc-notation \and
+
+Another neat example is the declaration of `\exists` that allows multiple binders:
+
 ```
+notation \exists X P = \neg (\forall X (\neg P)) [binder, right-assoc]
+```
+
+Here, the `right-assoc` attribute tells the verifier to accept patterns of the form `\exists X1 X2 ... Xn P` as a notation of the pattern `(\exists X1 ... (\exists Xn P) ...)`.
 
-Volatile `\forall` can be defined as
+___QUESTION.___ How to do the same thing for `\forall`? Note that unlike `\exists`, `\forall` is one of the primitive construct that takes exactly one binder, as specified by the AppKore syntax. Can we declare:
```

Comment:
Because `\forall` is built into the language, could we simply say, it is built in with the `right-assoc` attribute?

Reply:
I defined a new notation `\forall*` and `\exists*` which can take multiple arguments. `\forall` is not a notation but a builtin syntax of AppKore, so it feels strange if we "declare" it with `[right-assoc]`.